# SUMMER ANALYTICS 2022

## WEEK-4 ASSIGNMENT

In [2]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import matplotlib.pyplot as plt # data visualization
import seaborn as sns # statistical data visualization
from sklearn.preprocessing import LabelEncoder
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

**Let's explore the dataset shall we?**

The Car Evaluation Database contains examples with the structural information removed, i.e., directly relates CAR to the six input attributes: buying, maintenance cost, doors, persons, lug_boot, safety.
In this notebook we will go through an in depth analysis of sound and how we can classify and ultimately understand it.

I suggest you to use the internet before asking a doubt. Most of your doubts will disappear once you read the documentation or search StackOverflow!

Note: Pay close attention to the markdown cells and the comments.

In [3]:
# Read the csv file as data
df = pd.read_csv('car_evaluation.csv')
# you may print first few rows
df.head(1000)

vhigh vhigh.1  2   2.1  small   low  unacc
0    vhigh   vhigh  2     2  small   med  unacc
1    vhigh   vhigh  2     2  small  high  unacc
2    vhigh   vhigh  2     2    med   low  unacc
3    vhigh   vhigh  2     2    med   med  unacc
4    vhigh   vhigh  2     2    med  high  unacc
..     ...     ... ..   ...    ...   ...    ...
995    med    high  2  more    big   low  unacc
996    med    high  2  more    big   med    acc
997    med    high  2  more    big  high    acc
998    med    high  3     2  small   low  unacc
999    med    high  3     2  small   med  unacc

[1000 rows x 7 columns]

Attribute Information->

Class Values: unacc, acc, good, vgood

Attributes->
buying- vhigh, high, med, low.
maint- vhigh, high, med, low.
doors -2, 3, 4, 5more.
persons-2, 4, more.
lug_boot- small, med, big.
safety- low, med, high.

In [4]:
df.describe()

vhigh vhigh.1     2   2.1 small   low  unacc
count   1727    1727  1727  1727  1727  1727   1727
unique     4       4     4     3     3     3      4
top     high    high     3     4   med   med  unacc
freq     432     432   432   576   576   576   1209

In [5]:
# view dimensions of dataset
df.shape

(1727, 7)

We can see that there are 1727 instances and 7 variables in the data set.



**Rename column names**
We can see that the dataset does not have proper column names. The columns are merely labelled as 0,1,2.... and so on. We should give proper names to the columns. I will do it as follows:-

In [6]:
col_names = ['buying', 'maint', 'doors', 'persons', 'lug_boot', 'safety', 'class']
df.columns = col_names
col_names

['buying', 'maint', 'doors', 'persons', 'lug_boot', 'safety', 'class']

Frequency distribution of values in variables
Now, I will check the frequency counts of categorical variable

In [7]:
col_names = ['buying', 'maint', 'doors', 'persons', 'lug_boot', 'safety', 'class']
for col in col_names:
    print(df[col].value_counts())

high     432
med      432
low      432
vhigh    431
Name: buying, dtype: int64
high     432
med      432
low      432
vhigh    431
Name: maint, dtype: int64
3        432
4        432
5more    432
2        431
Name: doors, dtype: int64
4       576
more    576
2       575
Name: persons, dtype: int64
med      576
big      576
small    575
Name: lug_boot, dtype: int64
med     576
high    576
low     575
Name: safety, dtype: int64
unacc    1209
acc       384
good       69
vgood      65
Name: class, dtype: int64


We can see that the doors and persons are categorical in nature. So, I will treat them as categorical variables.

Summary of variables
There are 7 variables in the dataset. All the variables are of categorical data type.
These are given by buying, maint, doors, persons, lug_boot, safety and class.
class is the target variable.

#### Missing values in variables¶


In [8]:
# check missing values in variables
#write your code here
df.isnull().sum()

buying      0
maint       0
doors       0
persons     0
lug_boot    0
safety      0
class       0
dtype: int64

**Declare feature vector and target variable**

In [9]:
X = df.drop(['class'], axis=1)
y = df['class']

**Split data into separate training and test set**

In [10]:
#  split X and y into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.33, random_state = 42)

In [11]:
# check the shape of X_train and X_test
print(X_train.shape)
print(X_test.shape)

(1157, 6)
(570, 6)


Feature Engineering

Feature Engineering is the process of transforming raw data into useful features that help us to understand our model better and increase its predictive power. I will carry out feature engineering on different types of variables.

In [12]:
# check data types in X_train
X_train.dtypes

buying      object
maint       object
doors       object
persons     object
lug_boot    object
safety      object
dtype: object

Encode categorical variables
Now, I will encode the categorical variables.

In [13]:
X_train.head()

buying  maint  doors persons lug_boot safety
83    vhigh  vhigh  5more       2      med    low
48    vhigh  vhigh      3    more      med    med
468    high  vhigh      3       4    small    med
155   vhigh   high      3    more      med    low
1043    med   high      4    more    small    low

We can see that all the variables are ordinal categorical data type.



In [14]:
X_train = X_train.apply(LabelEncoder().fit_transform)
X_test = X_test.apply(LabelEncoder().fit_transform)

In [15]:
y_train = y_train.apply(LabelEncoder().fit_transform)
y_test = y_test.apply(LabelEncoder().fit_transform)

ValueError: y should be a 1d array, got an array of shape () instead.

In [ ]:
X_train.head()

In [ ]:
X_test.head()

Now let's perform Machine Learning Classification. We'll be predicting the genre based on the given features.

In [ ]:
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification

from xgboost import XGBClassifier
from xgboost import XGBRFClassifier
from xgboost import plot_tree, plot_importance

from sklearn.metrics import confusion_matrix, accuracy_score, roc_auc_score, roc_curve
from sklearn import preprocessing
from sklearn.feature_selection import RFE
from sklearn.ensemble import AdaBoostClassifier

Creating a Predefined function to assess the accuracy of a model. This will be the scoring function

In [ ]:
def score(model, title = "Default"):
    model.fit(X_train, y_train)
    preds = model.predict(X_test)
    print(preds)
#     print(confusion_matrix(y_test, preds))
    accuracy = round(accuracy_score(y_test, preds), 5)
    print('Accuracy for', title, ':', accuracy, '\n')

 Let's see how a basic Logistic Regressor works on this!

In [ ]:
lg = LogisticRegression(random_state=0, solver='lbfgs', multi_class='multinomial')
score(lg, "Logistic Regression")

**Now it's your turn. Train the following:**

4. **AdaBoostClassifier** with n_estimators=1000 & random_state=0
5. **RandomForestClassifier** with n_estimators=1000, max_depth=10, random_state=0

**IMPORTANT**: Use the default values for other hyper parameters apart from these. Do **NOT** change these values 

In [ ]:
# AdaBoostClassifier
Ada = AdaBoostClassifier(n_estimators=1000,random_state=0)
score(Ada,'AdaBoostClassifier')

In [ ]:
# Random Forest
rf = RandomForestClassifier(n_estimators=1000,max_depth=10,random_state=0)
score(rf,'Random Forest')

AdaBoost performs poorly, but Random Forest looks great!

Now look at the import statements. There may be a few models you may or may not have seen. Use those models to beat the score you achieved in the cell above! Play around with their hyper parameters.
We suggest going through their documentations before using them.

In [ ]:
# Naive Bayes: GaussianNB
nb = GaussianNB()
score(nb)
# KNN: KNeighborsClassifier
knn = KNeighborsClassifier()
score(knn)
# Decission trees: DecisionTreeClassifier
dt = DecisionTreeClassifier(criterion='gini')
score(dt)

Visualize decision-trees

In [ ]:
plt.figure(figsize=(200,30))

from sklearn import tree

tree.plot_tree(dt)

Decision Tree Classifier with criterion gini index,max_depth=10, random_state=0

In [ ]:
tree_gini = DecisionTreeClassifier(criterion='gini', max_depth=10, random_state=0)
score(tree_gini,"DecisionTreeClassifier")

Decision Tree Classifier with criterion entripy index,max_depth=8, random_state=0

In [ ]:
tree_entropy = DecisionTreeClassifier(criterion='entropy', max_depth=8, random_state=0)
score(tree_entropy,"DecisionTreeClassifier")

Use XGBClassifier with n_estimators=1000 and learning_rate=0.01

In [ ]:
# XGBClassifier
xgb = XGBClassifier(n_estimators=1000, learning_rate=0.01)
score(xgb, "XGBClassifier") 

Visualize XGBClassifier

In [ ]:
from xgboost import plot_tree
fig, ax = plt.subplots(figsize=(30, 30))
plot_tree(xgb, ax=ax)
plt.show()

Now, heres a task for you. Try XGBRFClassifier on your own.

In [ ]:
xgbr = XGBRFClassifier(n_estimators=1000, learning_rate=0.1)
score(xgbr)

Plot the Confusion Matrix for XGBClassifier with parameters n_estimators=1000, learning_rate=0.1

In [ ]:
# Cell to plot Confusion Matrix
xgb = XGBClassifier(n_estimators=1000, learning_rate=0.1)
xgb.fit(X_train, y_train)

preds = xgb.predict(X_test)
score(xgb, "XGBClassifier")

# Confusion Matrix
confusion_matr = confusion_matrix(y_test, preds) #normalize = 'true'
plt.figure(figsize = (16, 9))
sns.heatmap(confusion_matr, cmap="Blues", annot=True, 
            xticklabels = ["unacc","acc","good","vgood"],
           yticklabels=["unacc","acc","good","vgood"]);


8. Find the feature importance for XGBClassifier. 

*Hint: It's an inbuilt member variable*

In [ ]:
# Cell to find Feature Importance
pd.DataFrame({'Features':X.columns,
              "Importance":xgb.feature_importances_}).sort_values(by = ['Importance'],
                                                                  ascending=False)

**Classification Report**
Classification report is another way to evaluate the classification model performance. It displays the precision, recall, f1 and support scores for the model. I have described these terms in later.

We can print a classification for XGBClassifierwith parameters n_estimators=1000, learning_rate=0.1

In [ ]:
#YOUR CODE HERE9
from sklearn.metrics import classification_report
print(classification_report(y_test, preds))
